### PR TITLE
feat(router): deprecate preserveQueryParams, add queryParamsHandling.

### PR DESCRIPTION
**Pull request by dejour**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
#13806

**What is the new behavior?**
add queryParamsHandlling option

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
PR close #13825

### CODE REVIEW

**Comment by dejour** [comment]
@vicb pls review.

File: modules/@angular/router/src/router.ts [hunk 1]
```diff
@@ -463,11 +475,25 @@ export class Router {
...
   createUrlTree(
-      commands: any[], {relativeTo, queryParams, fragment, preserveQueryParams,
+      commands: any[], {relativeTo, queryParams, fragment, preserveQueryParams, queryParamsHandling,
                         preserveFragment}: NavigationExtras = {}): UrlTree {
     const a = relativeTo || this.routerState.root;
-    const q = preserveQueryParams ? this.currentUrlTree.queryParams : queryParams;
     const f = preserveFragment ? this.currentUrlTree.fragment : fragment;
+    let q: Params = null;
+    if (queryParamsHandling) {
+      switch (queryParamsHandling) {
```
Comment (DzmitryShylovich) [inline modules/@angular/router/src/router.ts]: `q = queryParamsHandling === 'merge' ? merge(this.currentUrlTree.queryParams, queryParams) : this.currentUrlTree.queryParams;`

Comment (dejour) [inline modules/@angular/router/src/router.ts]: it occurs to me that default case should be queryParams.so should it still write like this.

**Comment by paramsingh88** [comment]
@vicb Can we have it merged now? @dejour it fixes #13825 right?

File: modules/@angular/router/src/directives/router_link.ts [hunk 2]
```diff
@@ -59,12 +59,23 @@ import {UrlTree} from '../url_tree';
...
  *
+ * deprecated, use queryParamsHandling instead
+ *
  * ```
  * <a [routerLink]="['/user/bob']" preserveQueryParams preserveFragment>
  *   link to user component
  * </a>
  * ```
  *
+ * You can tell the directive to how to handle queryParams, available options are 'merge','preserve'
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_link.ts]: available options are:
- 'merge' merge the `queryParams` into the current qp,
- 'preserve' prserve the current qp,
- default / '' use the `queryParams` only.

File: modules/@angular/router/src/router.ts [hunk 3]
```diff
@@ -102,12 +102,24 @@ export interface NavigationExtras {
   /**
   * Preserves the query parameters for the next navigation.
   *
+  * @deprecated use queryParamsHandling instead
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: nits: move `@deprecated` just at the end of the comment, closer to the code

File: modules/@angular/router/src/router.ts [hunk 4]
```diff
@@ -102,12 +102,24 @@ export interface NavigationExtras {
...
   */
   preserveQueryParams?: boolean;
+
+  /**
+  *  config strategy to handle the query parameters for the next navigation.
+  *
+  * ```
+  * // from /results?page=1 to /view?page=1&page=2
+  * this.router.navigate(['/view'], { queryParams: { page: 2 },  queryParamsHandling: "merge" });
+  * ```
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: Document the options here, link to here from the directives ?

Also we should probably create a type `"merge"|"preserve"|""`

File: modules/@angular/router/src/router.ts [hunk 5]
```diff
@@ -463,11 +475,25 @@ export class Router {
...
+          q = merge(this.currentUrlTree.queryParams, queryParams);
+          break;
+        case 'preserve':
+          q = this.currentUrlTree.queryParams;
+          break;
+        default:
+          q = queryParams;
+      }
+    } else {
+      q = preserveQueryParams ? this.currentUrlTree.queryParams : queryParams;
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: add a deprecation warning in dev mode when `pqp` is used at the top of this function

File: modules/@angular/router/src/directives/router_link.ts [hunk 6]
```diff
@@ -59,12 +59,23 @@ import {UrlTree} from '../url_tree';
  *
  * You can also tell the directive to preserve the current query params and fragment:
  *
+ * deprecated, use queryParamsHandling instead
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_link.ts]: Please quote `qph`

File: modules/@angular/router/src/directives/router_link.ts [hunk 7]
```diff
@@ -84,6 +95,7 @@ export class RouterLink {
   @Input() queryParams: {[k: string]: any};
   @Input() fragment: string;
   @Input() preserveQueryParams: boolean;
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_link.ts]: please add a getter/setter and print a deprecation warning in dev mode

**Comment by vicb** [comment]
@dejour thanks for the updated PR. I have discussed it internally, should be merged when the feedback is addressed.

@paramsingh88 PRs are merged when they are reviewed and approved by the team.

File: modules/@angular/router/src/router.ts (hunk 4 above)
Comment (dejour) [inline modules/@angular/router/src/router.ts]: added link from directives

**Comment by dejour** [comment]
@vicb updated.

File: modules/@angular/router/src/directives/router_link.ts [hunk 8]
```diff
@@ -106,6 +122,14 @@ export class RouterLink {
     }
   }
 
+  @Input()
+  set preserveQueryParams(value: boolean) {
+    if (isDevMode()) {
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_link.ts]: please make sure `console.warn` is available

File: modules/@angular/router/src/directives/router_link.ts [hunk 9]
```diff
@@ -106,6 +122,14 @@ export class RouterLink {
...
+      console.warn('preserveQueryParams is deprecated!, use queryParamsHandling instead.');
+    }
+    this._preserveQueryParams = value;
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_link.ts]: use `queryParamsHandling` instead - yep there might be a conflict but there is a deprecation warning and both inputs should not be used at the same time

File: modules/@angular/router/src/directives/router_link.ts [hunk 10]
```diff
@@ -83,11 +98,12 @@ import {UrlTree} from '../url_tree';
 export class RouterLink {
   @Input() queryParams: {[k: string]: any};
   @Input() fragment: string;
-  @Input() preserveQueryParams: boolean;
+  @Input() queryParamsHandling: QueryParamsHandling;
   @Input() preserveFragment: boolean;
   @Input() skipLocationChange: boolean;
   @Input() replaceUrl: boolean;
   private commands: any[] = [];
+  private _preserveQueryParams: boolean;
```
Comment (vicb) [inline modules/@angular/router/src/directives/router_link.ts]: nit: remove the leading _ ti be consistent in this PR, thanks

**Comment by vicb** [comment]
A few last nits but mostly ok to go, thanks for the update !

**Comment by dejour** [comment]
@vicb done.

**Comment by vicb** [comment]
Thanks!

**Review by mhevery — APPROVED** [review]

**Pull request closed without merging**